model/protecode.py

```diff
@@ -71,3 +89,43 @@ def _required_attributes(self):
     def validate(self):
         # call credentials method to validate credentials
         self.credentials()
+
+
+def find_config(
+    group_id: int,
+    base_url: str,
+    cfg_factory=None,
+) -> ProtecodeConfig | None:
+    if not cfg_factory:
+        cfg_factory = ci.util.ctx().cfg_factory()
+
+    matching_cfgs = []
+
+    for cfg in cfg_factory._cfg_elements(cfg_type_name='protecode'):
+        cfg: ProtecodeConfig
+
+        if cfg.base_url() and not cfg.matches_base_url(base_url):
```

Comment:
mv this into `matches_base_url`. or rather: add a single method `matches`, like so (pseudo-code):

```
def _matches(self, base_url, group_id):
 score = 0
 if base_url:
   # todo: normalise urls (in particular `https://foo:443/bar` should match `https://foo/bar`)
   if base_url.startswith(self.base_url()):
     score += 1
   else:
     return -1 # scores < 0 indicate explicit mis-match
   if group_id:
     if group_id in self.group_ids():
       score += 1
     else:
       if not self.group_ids():
          pass # no score change if config does not specify a group-id
       return -1 # group_id did not match
   return score
```